Prioritize raising error message about unused parameters when rebuild_buckets fails

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45950 [NCCL] Provide additional information about NCCL error codes.
* #45942 Only populate grad accumulator to var mapping for find_unused_parameters=True in DDP
* **#45933 Prioritize raising error message about unused parameters when rebuild_buckets fails**

Occasionally users run DDP with models with unused params, in this
case we would like to surface an error message telling them to run with
find_unused_params=True. However, a recent change to rebuild_buckets logic (https://github.com/pytorch/pytorch/pull/44798) made
it so that we raise a size mismatch error when this happens, but the
information about unused parameters is likely to be more useful and likely to
be the most common case of failure. Prefer raising this error over the
subsequent size mismatch errors.

Differential Revision: [D24151256](https://our.internmc.facebook.com/intern/diff/D24151256/)

rohan-varma committed:
Prioritize raising error message about unused parameters when rebuild_buckets fails

Occasionally users run DDP with models with unused params, in this
case we would like to surface an error message telling them to run with
find_unused_params=True. However, a recent change to rebuild_buckets logic (https://github.com/pytorch/pytorch/pull/44798) made
it so that we raise a size mismatch error when this happens, but the
information about unused parameters is likely to be more useful and likely to
be the most common case of failure. Prefer raising this error over the
subsequent size mismatch errors.

Differential Revision: [D24151256](https://our.internmc.facebook.com/intern/diff/D24151256/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Prioritize raising error message about unused parameters when rebuild_buckets fails"


Occasionally users run DDP with models with unused params, in this
case we would like to surface an error message telling them to run with
find_unused_params=True. However, a recent change to rebuild_buckets logic (https://github.com/pytorch/pytorch/pull/44798) made
it so that we raise a size mismatch error when this happens, but the
information about unused parameters is likely to be more useful and likely to
be the most common case of failure. Prefer raising this error over the
subsequent size mismatch errors.

Differential Revision: [D24151256](https://our.internmc.facebook.com/intern/diff/D24151256/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -1330,6 +1309,12 @@ bool Reducer::rebuild_buckets() {
+  // has unused parameters for example, this will raise an error recommending to
+  // run with find_unused_parameters=True, instead of the size mismatch exception
+  // below.
+  ensure_prior_reduction_finished();
```
looks like `self.reducer._rebuild_buckets` is always invoked. Will it be sufficient if we move this to the front of the above if clause and then remove the other  `ensure_prior_reduction_finished` invocation in `prepare_for_backward`?

rohan-varma replied:
Yes, I believe this should be okay. Originally I was concerned about `require_backward_grad_sync` flag, but I don't think its an issue. If we don't `require_backward_grad_sync` this iteration, we should still have the previous iteration reduction finished, where "previous iteration" here means the previous iteration that required grad sync, or the initial state of `require_finalize_`, which is false.

mrshenli reviewed (approved):
LGTM!

rohan-varma committed:
Update on "Prioritize raising error message about unused parameters when rebuild_buckets fails"


Occasionally users run DDP with models with unused params, in this
case we would like to surface an error message telling them to run with
find_unused_params=True. However, a recent change to rebuild_buckets logic (https://github.com/pytorch/pytorch/pull/44798) made
it so that we raise a size mismatch error when this happens, but the
information about unused parameters is likely to be more useful and likely to
be the most common case of failure. Prefer raising this error over the
subsequent size mismatch errors.

Differential Revision: [D24151256](https://our.internmc.facebook.com/intern/diff/D24151256/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Prioritize raising error message about unused parameters when rebuild_buckets fails"


Occasionally users run DDP with models with unused params, in this
case we would like to surface an error message telling them to run with
find_unused_params=True. However, a recent change to rebuild_buckets logic (https://github.com/pytorch/pytorch/pull/44798) made
it so that we raise a size mismatch error when this happens, but the
information about unused parameters is likely to be more useful and likely to
be the most common case of failure. Prefer raising this error over the
subsequent size mismatch errors.

Differential Revision: [D24151256](https://our.internmc.facebook.com/intern/diff/D24151256/)

[ghstack-poisoned]
